module/zfs/spa.c

```diff
@@ -9589,6 +9602,164 @@ spa_sync_allpools(void)
+	kthread_t	*sta_thread;
+	kcondvar_t	sta_cv;
+	kmutex_t 	sta_lock;
+	int		sta_ready;
+} sync_tq_arg_t;
+
+/* unfortunately, taskqueues do not provide per-thread private data */
+static void
+spa_sync_tq_assign(void *arg)
+{
+	sync_tq_arg_t *sta = arg;
+
+	mutex_enter(&sta->sta_lock);
+	sta->sta_thread = curthread;
+	sta->sta_ready = 1;
+	cv_signal(&sta->sta_cv);
+	while (sta->sta_ready == 1)
+		cv_wait(&sta->sta_cv, &sta->sta_lock);
+	mutex_exit(&sta->sta_lock);
+}
+
+taskq_t *
+spa_sync_tq_create(spa_t *spa, const char *name)
+{
+	ASSERT(spa->spa_sync_tq == NULL);
+	ASSERT(spa->spa_alloc_count <= boot_ncpus);
+
+	/*
+	 * - do not allow more allocators than cpus.
+	 * - there may be more cpus than allocators.
+	 * - do not allow more sync taskq threads than allocators or cpus.
+	 */
+	int nthreads = spa->spa_alloc_count;
+
+	sync_tq_arg_t	*stq = kmem_zalloc(sizeof (*stq) * nthreads, KM_SLEEP);
+	spa->spa_syncthreads = kmem_zalloc(sizeof (spa_syncthread_info_t) *
+	    nthreads, KM_SLEEP);
+
+	spa->spa_sync_tq = taskq_create(name, nthreads, minclsyspri, nthreads,
+	    INT_MAX, TASKQ_PREPOPULATE);
+
+	/* spawn all syncthreads */
+	for (int i = 0; i < nthreads; i++) {
+		cv_init(&stq[i].sta_cv, NULL, CV_DEFAULT, NULL);
+		mutex_init(&stq[i].sta_lock, NULL, MUTEX_DEFAULT, NULL);
+		(void) taskq_dispatch(spa->spa_sync_tq, spa_sync_tq_assign,
+		    &stq[i], TQ_FRONT);
+	}
+
+	/* wait on all syncthreads to start */
+	for (int i = 0; i < nthreads; i++) {
+		mutex_enter(&stq[i].sta_lock);
+		while (stq[i].sta_ready == 0)
+			cv_wait(&stq[i].sta_cv, &stq[i].sta_lock);
+		mutex_exit(&stq[i].sta_lock);
+	}
+
+	/* let all syncthreads resume, finish */
+	for (int i = 0; i < nthreads; i++) {
+		mutex_enter(&stq[i].sta_lock);
+		stq[i].sta_ready = 2;
+		cv_broadcast(&stq[i].sta_cv);
+		mutex_exit(&stq[i].sta_lock);
+	}
+	taskq_wait(spa->spa_sync_tq);
+
+	spa_taskqs_t *tqs =
+	    &spa->spa_zio_taskq[ZIO_TYPE_WRITE][ZIO_TASKQ_ISSUE];
+
+	spa_syncthread_info_t *ti = spa->spa_syncthreads;
+	for (int i = 0, w = 0; i < nthreads; i++, w++, ti++) {
+		ti->sti_thread = stq[i].sta_thread;
+		if (w == tqs->stqs_count) {
+			w = 0;
+		}
+		ti->sti_wr_iss_tq = tqs->stqs_taskq[w];
+		mutex_destroy(&stq[i].sta_lock);
+		cv_destroy(&stq[i].sta_cv);
+	}
+	kmem_free(stq, sizeof (spa_syncthread_info_t) * nthreads);
+
+	VERIFY(spa->spa_alloc_count == nthreads);
+	ti = spa->spa_syncthreads;
+	for (int a = 0; a < nthreads; a++, ti++) {
+		if (ti->sti_num_alloc == 0)
+			ti->sti_first_alloc = a;
+		ti->sti_num_alloc++;
+	}
+	return (spa->spa_sync_tq);
+}
+
+void
+spa_sync_tq_destroy(spa_t *spa)
+{
+	ASSERT(spa->spa_sync_tq != NULL);
+
+	taskq_wait(spa->spa_sync_tq);
+	taskq_destroy(spa->spa_sync_tq);
+	kmem_free(spa->spa_syncthreads,
+	    sizeof (kthread_t *) * spa->spa_alloc_count);
+	spa->spa_sync_tq = NULL;
+}
+
+void
+spa_select_allocator(zio_t *zio)
+{
+	zbookmark_phys_t *bm = &zio->io_bookmark;
+	spa_t *spa = zio->io_spa;
+
+	VERIFY(zio->io_type == ZIO_TYPE_WRITE);
```

Comment:
Here and in other places I would prefer ASSERT() to be used for things that should never happen algorithmically, that we should be able to catch in CI or other debug builds.  While VERIFY() for things that we are less certain about in runtime, such as some errors that we may not handle, but that may be difficult to reproduce/test.  We do not want extra branching all over the production code.

Reply:
sounds good, will do

Reply:
updated, thanks